### PR TITLE
Adjust read capacity by wanted time

### PR DESCRIPTION
**Pull request by oh14**
Fix SSE option: add variable to use AES256 or aws:kms
Apply fix from markitx#63
New option to create backup in specified during time.
New option to use strongly consistent reads

### CODE REVIEW

File: lib/dynamo-backup.js [hunk 1]
```diff
@@ -164,43 +166,122 @@ DynamoBackup.prototype._getBackupPath = function () {
 DynamoBackup.prototype._copyTable = function (tableName, itemsReceived, callback) {
     var self = this;
     var ddb = new AWS.DynamoDB();
+    var startDescTime = new Date().getTime();
     ddb.describeTable({ TableName: tableName }, function (err, data) {
         if (err) {
             return callback(err);
         }
+        var descTime = new Date().getTime() - startDescTime - 0.3;
```
Comment (ChristopheBougere) [inline lib/dynamo-backup.js]: `- 0.3`?

File: lib/dynamo-backup.js [hunk 2]
```diff
@@ -16,7 +16,14 @@ function DynamoBackup(options) {
     options = options || {};
     this.excludedTables = options.excludedTables || [];
     this.includedTables = options.includedTables;
+    this.tablesToBackup = [];
+    this.consistentRead = !!options.consistentRead; // strongly consistent
+    this.consistentRatio = this.consistentRead ? 1 : 0.5; // eventually consistent consumes 1/2 of a read capacity unit.
+    this.singleCapacitySize = 4096 / this.consistentRatio; // 4KB for consistent, 8KB for eventually consistent
+    this.maxSingleScanReadCapacity = 1048576 / this.singleCapacitySize; // 1MB per scan page
```
Comment (ChristopheBougere) [inline lib/dynamo-backup.js]: `4096` and `1048576`could be stored in constants to be more explicit

File: lib/dynamo-backup.js [hunk 3]
```diff
@@ -164,43 +166,122 @@ DynamoBackup.prototype._getBackupPath = function () {
...
+        var originalReadCapacity = Table.ProvisionedThroughput.ReadCapacityUnits;
+        var originalWriteCapacity = Table.ProvisionedThroughput.WriteCapacityUnits; // needed for future update
         var readPercentage = self.readPercentage;
-        var limit = Math.max((data.Table.ProvisionedThroughput.ReadCapacityUnits * readPercentage) | 0, 1);
+        var maxCapacityUnits = Math.max(Math.floor(originalReadCapacity * readPercentage), 1);
+        var tableCapacityConsumption = Table.TableSizeBytes / self.singleCapacitySize;
+        var averageItemSize = Math.ceil(Table.TableSizeBytes / Table.ItemCount);
+
+        var minimumNbScans = Math.ceil(Table.TableSizeBytes / 1048576); // 1 scan can't return more than 1MB of data
+        var estimatedWriteTime = minimumNbScans * 0.035 * self.tablesToBackup.length; // ~35ms for 1MB response (estimation) with tables concurrency
```
Comment (ChristopheBougere) [inline lib/dynamo-backup.js]: `1048576` and `0.035` in constants

File: lib/dynamo-backup.js [hunk 4]
```diff
@@ -164,43 +166,122 @@ DynamoBackup.prototype._getBackupPath = function () {
...
+
+        // estimate the backup duration with the current read capacity
+        var nbSecondsToBackup = Math.ceil(tableCapacityConsumption / maxCapacityUnits) + estimatedWriteTime;
+
+        if( nbSecondsToBackup <= self.maxSecondsPerTable ) {
+            self._streamItems(tableName, null, 0, maxCapacityUnits, averageItemSize, itemsReceived, callback);
+            return;
+        }
+
+        var estimatedFetchRatio = speed / 16384 ; // estimation ratio: if >1: faster, <1: slower. (16KB = highWaterMark of stream)
```
Comment (ChristopheBougere) [inline lib/dynamo-backup.js]: constant too

File: lib/dynamo-backup.js [hunk 5]
```diff
@@ -269,4 +350,67 @@ DynamoBackup.prototype._getDataPipelineAttributeValueKey = function (type) {
...
+    var params = {
+        TableName: tableName,
+        ProvisionedThroughput: {
+            ReadCapacityUnits: newReadCapacity,
+            WriteCapacityUnits: newWriteCapacity
+        }
+    };
+
+    var ddb = new AWS.DynamoDB();
+    ddb.updateTable(params, function(err, data) {
```
Comment (ChristopheBougere) [inline lib/dynamo-backup.js]: =>
`ddb.updateTable(params, callback);

File: lib/dynamo-backup.js (hunk 1 above)
Comment (oh14) [reply lib/dynamo-backup.js]: time is in milliseconds, 0.3ms is the during time to parse response result (average)

**Review by ChristopheBougere — APPROVED** [review]
👍

**Pull request merged**